### ISSUE TITLE
Allow more than 20 tabs per set

### ISSUE THREAD

**Issue by abhibeckert** [brave/browser-laptop#6692]
Related to: #4986 Have an option to allow users to have as many tabs as needed (without using tabs group)

While waiting for a long term fix to the issue above, I'd like to see a quick interim fix by increasing the maximum number of tabs per set.

Commenting on #4986 @bradleyrichter said the main goal is to prevent tabs becoming "unrecognizable nubs". I have a larger screen than most people, so for me with 20 tabs everything is perfectly usable with the first 3 or 4 words of the page title still visible.

For me, a tab is "unusable" when the tabs are no-longer wide enough to display the favicon. You can fit 100 tabs on a common 2560px display, or up to 250 tabs on the largest/most expensive displays, without making tabs "unrecognizable nubs".

So, in summary, I think Brave should:

 * allow up to 200 tabs per set
 * improve drawing at small sizes, favicon should be centered and not cropped: 
![screen shot 2017-01-17 at 1 23 46 pm](https://cloud.githubusercontent.com/assets/266088/22006603/3affa048-dcb8-11e6-941f-f0acf15af85e.png)
 * find a better long term solution as part of #4986

**Comment by bsclifton** [comment]
This may get fixed with https://github.com/brave/browser-laptop/pull/5918

cc: @cezaraugusto

**Comment by cezaraugusto** [comment]
@abhibeckert @bsclifton yes weird tab's UI will be fixed after that.

About more than 20 tabs per page, we do have a planned feature, similar as requested, described on #955. It would be a mix of what we have currently and what Firefox does for scrolling tabs. Thoughts are more than welcome.

**Comment by abhibeckert** [comment]
@cezaraugusto my feeling there should be a user configurable minimum tab width, which might be "favicon only" or "favicon + a few words of the title".

Then, if your tabs don't fit at the current window width, the tab bar should simply be wider than the window and people can use a mouse/trackpad to scroll horizontally (as a Mac user, both my mouse and trackpad support horizontal scrolling, perhaps this is less common on Windows?).

I don't think grouping into sets is workable in any way, I should never be in a situation where I right click a link to open a tab and it opens in a different tab set. I need those two tabs visible at the same time.

For the interface, I'm a fan of how Safari does it:

<img width="436" alt="screen shot 2017-01-18 at 13 03 45" src="https://cloud.githubusercontent.com/assets/266088/22049386/957792c6-dd7e-11e6-8a16-dc755918d361.png">

The tabs on one (or both) sides get progressively smaller and the titles are cropped, making it clear you've got more visible that can't be seen.

Clicking on one of those smaller tabs activates it and also scrolls over by an entire "page". So if you click on a tab on the far left of the window, it becomes the active tab and the tab bar is scrolled so that it is on the far right of the window (if there are enough tabs to do that).

This means clicking on the left most tab is equivalent to clicking the left arrow to jump over to the next tab set in Brave.

**Comment by bradleyrichter** [comment]
We should be able to solve the two-tabs set-split problem with the fix @cezaraugusto mentions above. There is an interesting approach that we could work on which is to let the selected tab have a minimum and it's neighbors can reduce to the favicon.

With Safari's approach, I have no idea what a small tab contains until I click on it. So it becomes a cat and mouse game. "Is that it? Nope. Is that it? Nope!"  :   )

I do think sometimes the best solution seems to be the one you have become used to over time. After getting used to the Brave tab sets with the tab ear previews, I can't use anything else. 

We certainly need to keep refining so thanks for the great feedback.

**Comment by abhibeckert** [comment]
> With Safari's approach, I have no idea what a small tab contains until I click on it. So it becomes a cat and mouse game. "Is that it? Nope. Is that it? Nope!" : )
Safari has a minimum tab width of 120px. Here's the above screenshot without cropping to fit nicely on GitHub:

![screen shot 2017-01-18 at 16 16 05](https://cloud.githubusercontent.com/assets/266088/22052993/93162004-dd99-11e6-94cb-6e65951d8019.png)

Only a few are drawn smaller as an intuitive clue that there are more tabs to scroll over and see. It's an elegant design in my opinion (aside from the fact that I think the minimum tab width should be the width of a favicon).

> I do think sometimes the best solution seems to be the one you have become used to over time. 

I will never get used to having most of my tabs inaccessible without switching to another tab group.

**Comment by bradleyrichter** [comment]
Wow. That is quite a screen shot.

We will improve that use case for sure. I can see why the current tab sets are not flexible enough for this.




I think 100 per set, plus the new scrolling approach will bring it to a similar place as Safari in this regard.

**Comment by bsclifton** [comment]
@abhibeckert @bradleyrichter I have a proposal that I would like your thoughts on (similar to https://github.com/brave/browser-laptop/issues/6272)

I propose we can offer a new preference for the minimum tab size (maybe it can default to something like 64px).  After enabling, you could then add as many tabs as you'd like, until you hit the max width (they would then overflow to the next tab page).  If you resize the window, we'd need to consider the same logic; adding or removing tabs until the max width is hit (with regard to user specified min-tab-width).

**Comment by bradleyrichter** [comment]
@bsclifton This sounds like a good solution. We need to address the FS view problem that @abhibeckert has and this would work.

When we start extending the power of tab pages, we may need to work around this and allow a fixed number in a tab page/set. I want to be able to use tab pages for auto-grouping, manual grouping, floating into a new window, etc.

Here is a little mockup for a settings menu:
![image](https://cloud.githubusercontent.com/assets/13509546/22477356/90f23c08-e79a-11e6-955f-c5cde7e0d4e6.png)

**Comment by bsclifton** [comment]
@bradleyrichter this is wonderful 😄  those sizes line up nicely with the ones that @cezaraugusto has introduced with https://github.com/brave/browser-laptop/pull/6900

**Comment by bradleyrichter** [comment]
well, there is one important difference. We still have a fixed number of tabs per tab page. In order to do this, we need to implement variable N per set.

**Comment by luixxiul** [comment]
@bradleyrichter if the mockup will be implemented please update the mockup on #2137, https://github.com/brave/browser-laptop/issues/2137#issuecomment-276525852 too.
